Package manager consumer refresh

OliverEGreen opened the pull request:
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

(FILL ME IN) This section describes why this PR is here. Usually it would include a reference 
to the tracking task that it is part or all of the solution for.

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -21,8 +21,19 @@ public class PackageManagerSearchElementViewModel : BrowserItemViewModel, IEquat
         public ICommand DownloadLatestToCustomPathCommand { get; set; }
 
+        /// <summary>
+        /// A function used to determine whether the used has installed this package already.
```
`user`?

OliverEGreen replied:
:D

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -21,8 +21,19 @@ public class PackageManagerSearchElementViewModel : BrowserItemViewModel, IEquat
+        /// <summary>
+        /// Determines whether the user can click on the 'Install' button of a package.
+        /// </summary>
+        public bool IsInstallButtonEnabled => !Model.IsDeprecated && !IsPackageInstalledProperty;
```
Is the InstallButton not using a command?

OliverEGreen replied:
As in does it have a CanInstallPackage Boolean value already? 

It seems to use some freaky anonymous method call: 
`this.DownloadLatestCommand = new DelegateCommand(() => OnRequestDownload(Model.Header.versions.Last(), false));`

SHKnudsen replied:
Cool, maybe lets just add this check there then.

belive something like this should work:
```csharp
this.DownloadLatestCommand = new DelegateCommand(
    () => OnRequestDownload(Model.Header.versions.Last(), false), 
    () => Model.IsDeprecated && !IsPackageInstalledProperty);

OliverEGreen replied:
Should I remove the property? It's being bound to in the view in a few places as well so ideally not :)

OliverEGreen replied:
Actually I could go ahead and update those styles to use IsEnabled I guess....

OliverEGreen replied:
Can't do it, there's stuff outside the button (adjacent to it) which is gonna get real messy if I remove this property.

SHKnudsen replied:
But couldnt you just bind those other things to `IsInstalled`?

OliverEGreen replied:
true, a MultiBinding should do it!

OliverEGreen replied:
This means IsInstalled needs to be public btw, since we're binding to it

OliverEGreen replied:
Done

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -41,6 +52,45 @@ public PackageManagerSearchElementViewModel(PackageManagerSearchElement element,
+        /// assist communication between the PackageManagerSearchViewModel
+        /// and the PackageManagerSearchElementViewModel.
+        /// </summary>
+        /// <param name="element"></param>
+        /// <param name="canLogin"></param>
+        /// <param name="isPackageInstalledFunction"></param>
```
Add description to these

OliverEGreen replied:
Added

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -41,6 +52,45 @@ public PackageManagerSearchElementViewModel(PackageManagerSearchElement element,
+        /// </summary>
+        /// <param name="packageName"></param>
+        /// <returns></returns>
+        internal bool IsPackageInstalled(string packageName)
+        {
+            IsPackageInstalledProperty = isPackageInstalled(packageName);
+            return IsPackageInstalledProperty;
```
Thinking about this one, we should have access to the package name from the viewmodel here, right?

If we do, we can remove the packageName argument and just get it directly from the model

OliverEGreen replied:
I think the reason we had to it in this complex way is we have no access to the PackageSearchManagerViewModel (or Model) from within this class.

OliverEGreen replied:
Or do you mean the PackageSearchElementModel? That has the name

OliverEGreen replied:
Yeah good point, done!

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -41,6 +52,45 @@ public PackageManagerSearchElementViewModel(PackageManagerSearchElement element,
+        /// <summary>
+        /// A Boolean flag reporting whether or not the user already has this SearchElement's package installed.
+        /// </summary>
+        public bool IsPackageInstalledProperty
```
Don't really like that this is called `IsPackageInstalledProperty` can we rename to `IsInstalled`

OliverEGreen replied:
DOne

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -87,11 +137,23 @@ private static void GoToUrl(string url)
         }
 
         private List<String> CustomPackageFolders;
+        private bool isPackageInstalledProperty;
```
maybe move this up to the Property

OliverEGreen replied:
Done

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -87,11 +137,23 @@ private static void GoToUrl(string url)
+        /// <summary>
+        /// Checks if the user has installed a package with a given name.
+        /// </summary>
+        public void OnCheckIfPackageInstalled()
```
When is this getting called?

OliverEGreen replied:
Never, deleted

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs:
```diff
@@ -333,6 +342,24 @@ public PackageSearchState SearchState
+        /// <summary>
+        /// Opens the Package Details ViewExtension
+        /// </summary>
+        public DelegateCommand<object> ViewPackageDetailsCommand { get; set; }
```
Guess this currently dosent do anything?
Might be good to note that

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs:
```diff
@@ -451,6 +481,42 @@ public void SetSortingDirection(object sortingDir)
+        /// <param name="obj"></param>
+        public void ViewPackageDetails(object obj)
+        {
+            // to be implemented in future PR   
```
Ah!

SHKnudsen commented on src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml.cs:
```diff
@@ -14,9 +17,12 @@ namespace Dynamo.PackageManager.UI
     /// </summary>
     public partial class PackageManagerSearchView : Window
     {
+        public PackageManagerSearchViewModel PackageManagerSearchViewModel { get;  }
```
consider just calling this `ViewModel` or something, it can get confusing when Properties has the same name as the type

OliverEGreen replied:
Updated! 

I actually find it easiest when the type name and object match 👁️

SHKnudsen replied:
Maybe thats actually fine... 
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members?redirectedfrom=MSDN#names-of-properties

I only mentioned it because i remember the Dynamo team commented on it before

SHKnudsen replied:
Also, does this need to be Public?

SHKnudsen commented on src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml.cs:
```diff
@@ -123,5 +129,54 @@ private void OnRequestShowFileDialog(object sender, PackagePathEventArgs e)
+        /// <param name="e"></param>
+        private void ViewDetailsButton_OnClick(object sender, RoutedEventArgs e)
+        {
+            if (!(this.DataContext is PackageManagerSearchViewModel packageManagerSearchViewModel)) return;
```
how come we dont just use the `ViewModel` property here?

OliverEGreen replied:
Updated

SHKnudsen commented on src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml.cs:
```diff
@@ -123,5 +129,54 @@ private void OnRequestShowFileDialog(object sender, PackagePathEventArgs e)
+        {
+            Button button = sender as Button;
+            
+            if (!(button.DataContext is PackageDownloadHandle packageDownloadHandle)) return;
```
same question as above

OliverEGreen replied:
On line 179 we do get the ViewModel, but we want access to the PackageDownloadHandle, which comes easiest from the button we're clicking. It's a ListView, so we want the DataContext of the item in the collection. Then we pass that to the command on the VM.

SHKnudsen replied:
Ah i see, i think i just read this to fast and didnt notice this was something else, my bad 🙃

SHKnudsen commented on src/DynamoPackages/PackageManagerSearchElement.cs:
```diff
@@ -42,7 +39,7 @@ public int Votes
         public string LatestVersion { get { return Header.versions[Header.versions.Count - 1].version; } }
         public string LatestVersionCreated { get { return Header.versions[Header.versions.Count - 1].created; } }
-
+        
```
seems like there arent really any changes in this class, so lets just revert it back

OliverEGreen replied:
Done

OliverEGreen commented on src/DynamoCoreWpf/UI/Converters.cs:
```diff
@@ -113,22 +114,13 @@ public class DependencyListToStringConverter : IValueConverter
-                List<string> depList = (List<string>)value;
-                foreach (var dep in depList)
-                {
-                    if (PythonEngineList.IndexOf(dep) != -1)
```
Good spot! Removed

OliverEGreen commented on src/DynamoCoreWpf/DynamoCoreWpf.csproj:
```diff
@@ -84,6 +84,7 @@
       <HintPath>..\packages\SharpDX.4.2.0\lib\net45\SharpDX.dll</HintPath>
     </Reference>
     <Reference Include="SharpDX.D3DCompiler">
+      <Private>True</Private>
```
Reset

OliverEGreen commented on src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs:
```diff
@@ -64,4 +64,4 @@
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyFileVersion("2.13.0.2443")]
+[assembly: AssemblyFileVersion("2.13.0.2374")]
```
Ah damn I manually turned this one back! Didn't realised it had jumped ahead.

OliverEGreen commented on src/DynamoCoreWpf/UI/Converters.cs:
```diff
@@ -312,6 +304,26 @@ public class NonEmptyStringToCollapsedConverter : IValueConverter
+            CultureInfo culture)
+        {
+            if (!(value is bool booleanValue)) return null;
+            return booleanValue ? "Installed" : "Install";
```
Thought I'd got all of these! Added

OliverEGreen commented on src/DynamoCoreWpf/UI/Converters.cs:
```diff
@@ -1182,6 +1194,58 @@ public object ConvertBack(object value, Type targetType, object parameter, Syste
     }
 
+    /// <summary>
+    /// Used in the Dynamo package manager to hide or show a label next to the package's name.
```
Added more explanation

SHKnudsen commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:
```diff
@@ -41,6 +52,46 @@ public PackageManagerSearchElementViewModel(PackageManagerSearchElement element,
+        public bool IsInstalled
+        {
+            get => isInstalled;
+            set
```
It might make sense the we make the setter for this private, we only really want `IsPackageInstalled()` to update this

OliverEGreen replied:
Done

OliverEGreen replied:
Had to undo this since we're binding to it in the view now (see above)

SHKnudsen reviewed (changes requested):
One last comment + one additional response on previous comment after that LGTM

SHKnudsen commented:
@OliverEGreen just noticed we still have these
![image](https://user-images.githubusercontent.com/13732445/132842952-702eb2cd-4228-4c49-b47a-17c8d35b2230.png)

Are they actually needed?

OliverEGreen commented:
> @OliverEGreen just noticed we still have these
> ![image](https://user-images.githubusercontent.com/13732445/132842952-702eb2cd-4228-4c49-b47a-17c8d35b2230.png)
> 
> Are they actually needed?

They are, yeah! I've tried removing them and it won't compile. I think they get set in that new constructor we made.

SHKnudsen commented:
> > @OliverEGreen just noticed we still have these
> > ![image](https://user-images.githubusercontent.com/13732445/132842952-702eb2cd-4228-4c49-b47a-17c8d35b2230.png)
> > Are they actually needed?
> 
> They are, yeah! I've tried removing them and it won't compile. I think they get set in that new constructor we made.

Hm, where are they used? I dont think we are actually invoking that event anywhere, think we removed that?

OliverEGreen commented:
> > > @OliverEGreen just noticed we still have these
> > > ![image](https://user-images.githubusercontent.com/13732445/132842952-702eb2cd-4228-4c49-b47a-17c8d35b2230.png)
> > > Are they actually needed?
> > 
> > 
> > They are, yeah! I've tried removing them and it won't compile. I think they get set in that new constructor we made.
> 
> Hm, where are they used? I dont think we are actually invoking that event anywhere, think we removed that?

PackageManagerSearchViewModel line 642:

`   element.CheckIfPackageInstalled += this.UserAlreadyHasPackageInstalled;`

SHKnudsen commented:
> > > > @OliverEGreen just noticed we still have these
> > > > ![image](https://user-images.githubusercontent.com/13732445/132842952-702eb2cd-4228-4c49-b47a-17c8d35b2230.png)
> > > > Are they actually needed?
> > > 
> > > 
> > > They are, yeah! I've tried removing them and it won't compile. I think they get set in that new constructor we made.
> > 
> > 
> > Hm, where are they used? I dont think we are actually invoking that event anywhere, think we removed that?
> 
> PackageManagerSearchViewModel line 642:
> 
> ` element.CheckIfPackageInstalled += this.UserAlreadyHasPackageInstalled;`

Arh yes, was accidentally searching for the type....

But same question then, are we actually invoking this event anywhere?

SHKnudsen commented:
Maybe the invoke shouldn't have been removed?

OliverEGreen commented:
> > > > > @OliverEGreen just noticed we still have these
> > > > > ![image](https://user-images.githubusercontent.com/13732445/132842952-702eb2cd-4228-4c49-b47a-17c8d35b2230.png)
> > > > > Are they actually needed?
> > > > 
> > > > 
> > > > They are, yeah! I've tried removing them and it won't compile. I think they get set in that new constructor we made.
> > > 
> > > 
> > > Hm, where are they used? I dont think we are actually invoking that event anywhere, think we removed that?
> > 
> > 
> > PackageManagerSearchViewModel line 642:
> > ` element.CheckIfPackageInstalled += this.UserAlreadyHasPackageInstalled;`
> 
> Arh yes, was accidentally searching for the type....
> 
> But same question then, are we actually invoking this event anywhere?

Hmm... nope seems not! :D 
Removed them.
